keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,997 @@
+```
+
+### Test Plan
+
+The new code will be covered by unit tests that execute as part of `pull-kubernetes-unit`.
+
+Converted components will be tested by exercising them with JSON output at high
+log levels to emit as many log messages as possible. Analysis of those logs
+will detect duplicate keys that might occur when a caller uses `WithValues` and
+a callee adds the same value in a log message.
+
+Static code analysis cannot detect this, or at least not easily. What it can
+check is that the individual log calls pass valid key/value pairs.
+
+
+### Graduation Criteria
+
+#### Alpha
+
+- Common utility code available (logcheck with the additional checks,
+  `k8s.io/klogr` v0.1.0)
+- Documentation for developers available at https://github.com/kubernetes/community
+- At least kube-scheduler framework and some plugins converted
+- Initial e2e tests completed and enabled in Prow
+
+#### Beta
+
+- All of kube-scheduler (in-tree) and CSI external-provisioner (out-of-tree) converted
+- Users of klog notified through a Kubernetes blog post and an email to
+  dev@kubernetes.io that a logger must be set with k8s.io/klogr.
+- Issues filed against kubernetes-sigs projects with a request to set a logger
+  with k8s.io/klogr. These issues will have a link to the enhancement issue
+  (for cross-referencing) and a link to the developer documentation.
+
+#### GA
+
+- All code in kubernetes/kubernetes converted to contextual logging,
+  no dependency on klog anymore
+- All code under kubernetes-sigs sets a logger with k8s.io/klogr
+  and therefore is ready for the next k/k release.
+- Gather feedback from developers and surveys
+- Allowing time for feedback
+
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+
+### Upgrade / Downgrade Strategy
+
+Not applicable. The log output will be determined by what is implemented in the
+code that currently runs.
+
+### Version Skew Strategy
+
+Not applicable. The log output format is the same as before, therefore other
+components are not affected.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+This is not a feature in the traditional sense. Converted log calls will always
+use contextual logging, the only supported rollback is to downgrade to a
+previous release.
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Other
+  - Describe the mechanism: updating to a new release where code was converted.
+  - Will enabling / disabling the feature require downtime of the control
+    plane? Yes, because components must be restarted.
+  - Will enabling / disabling the feature require downtime or reprovisioning of
+    a node? kubelet must be restarted, but the node itself can continue to run.
```

Comment:
This is not a mechanism to enable/disable the feature.

Would it be possible to add a feature flag that, if disabled, used the global klog context, and when enabled, used contextual logging?

Reply:
Yes, see above. I actually wrote that before seeing this question.